File: server_code/LoRa.py
import serial
import time
from typing import Literal, List
from enum import Enum
import re

import logging
logger = logging.getLogger('[LoRa]')

LORA_MAX_SIZE = 255 # max packet size

class DataRate(Enum):
    SF7 = 7
    SF8 = 8
    SF9 = 9
    SF10 = 10
    SF11 = 11
    SF12 = 12

class Bandwidth(Enum):
    BANDWIDTH_7_81_KHZ = 0
    BANDWIDTH_15_63_KHZ = 1
    BANDWIDTH_31_25_KHZ = 2
    BANDWIDTH_62_5_KHZ = 3
    BANDWIDTH_125_KHZ = 4
    BANDWIDTH_250_KHZ = 5
    BANDWIDTH_500_KHZ = 6

class CodeRate(Enum):
    RATE_4_BY_5 = 5
    RATE_4_BY_6 = 6
    RATE_4_BY_7 = 7
    RATE_4_BY_8 = 8

class LNA(Enum):
    OFF = 0
    ON = 1

class LowDrOpt(Enum):
    OFF = 0
    ON = 1
    AUTO = 2

class LoRaState(Enum):
    RX = 0
    TX = 1
    IDLE = 2
    SLEEP = 3 # same as idle but backwards compatible with RA-08H

AT_CCONF: bytes = b"AT+CCONF"
AT_CTX: bytes = b"AT+CTX"
AT_CRX: bytes = b"AT+CRX"
AT_COFF: bytes = b"AT+COFF"
AT_RESET: bytes = b"ATZ"

def hex_char_to_byte(c: int) -> int:
    if c >= ord('0') and c <= ord('9'):
        return c - ord('0')
    elif c >= ord('a') and c <= ord('f'):
        return c - ord('a') + 10
    elif c >= ord('A') and c <= ord('F'):
        return c - ord('A') + 10
    # invalid
    return -1

class LoRa:
    def __init__(self, lora_serial: serial.Serial, timeout: float, num_retries: int, freq: int, data_rate: DataRate, bandwidth: Bandwidth, code_rate: CodeRate, tx_power: int, lna: LNA, low_dr_opt: LowDrOpt) -> None:
        self._lora_serial: serial.Serial = lora_serial
        self._serial_buffer: bytes = b""
        self._timeout: float = timeout
        self._num_retries: int = num_retries
        self._freq: int = freq
        self._data_rate: DataRate = data_rate
        self._bandwidth: Bandwidth = bandwidth
        self._code_rate: CodeRate = code_rate
        self._tx_power: int = tx_power
        self._lna: LNA = lna
        self._low_dr_opt: LowDrOpt = low_dr_opt
        self._state: LoRaState = LoRaState.IDLE
        self._buffer: bytes = b""
        self._serial_buffer: bytes = b""

    def begin(self) -> bool:
        self._flush_serial()
        if not self.reset():
            return False
        if not self.configure():
            return False
        self._state = LoRaState.IDLE
        return True
    
    def reset(self) -> bool:
        command_buffer: bytes = AT_RESET + b"\r\n"
        expected_buffer: bytes = b"AT? to list all available functions\r\n"
        return self._send_command(command_buffer, expected_buffer)

    def configure(self) -> bool:
        command_buffer = AT_CCONF + f"={self._freq}:{self._tx_power}:{self._bandwidth.value}:{self._data_rate.value}:4/{self._code_rate.value}:{self._lna.value}:{self._low_dr_opt.value}\r\n".encode()
        expected_buffer = b"\r\nOK\r\n"
        return self._send_command(command_buffer, expected_buffer)

    def send(self, data: bytes | List[int]) -> bool:
        if len(data) > LORA_MAX_SIZE:
            return False
        
        SAVED_STATE: LoRaState = self._state
        if not self.set_state(LoRaState.TX):
            self.set_state(SAVED_STATE)
            return False
        
        self._buffer = AT_CTX + b"="
        # append data in 2-byte padded hex format with no separation
        for datum in data:
            self._buffer += f"{datum:02x}".encode()
        
        self._buffer += b"\r\n"

        EXPECTED_BUFER: bytes = b"OnTxDone\r\n"
        if not self._send_command(self._buffer, EXPECTED_BUFER):
            self.set_state(SAVED_STATE)
            return False
        
        self.set_state(SAVED_STATE) # ignore return value, we still successfully sent
        return True

    def set_state(self, updated_state: LoRaState) -> bool:
        if updated_state == LoRaState.RX:
            return self.enter_rx_mode_()
        elif updated_state == LoRaState.TX:
            return self.enter_tx_mode_()
        elif updated_state == LoRaState.IDLE:
            return self.enter_idle_mode_()
        elif updated_state == LoRaState.SLEEP:
            return self.enter_idle_mode_() # `IDLE` is already low power
        else:
            return False

    def get_state(self) -> LoRaState:
        return self._state
    
    def enter_rx_mode_(self) -> bool:
        if not self.configure():
            return False
        command_buffer: bytes = AT_CRX + b"\r\n"
        expected_buffer: bytes = b"Receiving...\r\n"
        return self._send_command(command_buffer, expected_buffer)
    
    def enter_tx_mode_(self) -> bool:
        if not self.enter_idle_mode_():
            return False
        return self.configure()

    def enter_idle_mode_(self) -> bool:
        if not self.configure():
            return False
        command_buffer: bytes = AT_COFF + b"\r\n"
        expected_buffer: bytes = b"Idle...\r\n"
        return self._send_command(command_buffer, expected_buffer)

    def recv(self) -> bool:
        SAVED_STATE: LoRaState = self._state
        def success() -> bool:
            self.set_state(SAVED_STATE)
            return True
        def fail() -> bool:
            self.set_state(SAVED_STATE)
            return False
        if not self.set_state(LoRaState.RX):
            return fail()
        for i in range(self._num_retries):
            self._buffer = b""
            expected_buffer: bytes = b"Recv:\r\n"
            if not self._expected(expected_buffer):
                continue
            c1: int = self._read_blocking()
            if c1 == -1:
                return fail()
            c2: int = self._read_blocking()
            if c2 == -1:
                return fail()
            while c1 != ord('\r'):
                if len(self._buffer) >= LORA_MAX_SIZE:
                    return fail()
                c1_byte = hex_char_to_byte(c1)
                c2_byte = hex_char_to_byte(c2)
                if (c1_byte == -1) or (c2_byte == -1): # invalid hex formats
                    return fail()
                self._buffer += ((c1_byte << 4) | c2_byte).to_bytes(1, byteorder='little')
                if self._read_blocking() != ord(' '): # flush ' ' char
                    return fail()
                c1 = self._read_blocking()
                if c1 == -1:
                    return fail()
                c2 = self._read_blocking()
                if c2 == -1:
                    return fail()
            # the rx ends with `"Data end\r\nrssi = {rssi} dBm, snr = {snr} dB\r\n"`
            expected_buffer = b"Data end\r\nrssi = "
            if not self._expected(expected_buffer):
                return fail()
            expected_buffer = b" dB\r\n"
            if not self._expected(expected_buffer):
                return fail()
            
            regex_match = re.match(rb"([+-]?\d+) dBm, snr = ([+-]?\d+) dB", self._serial_buffer)
            if regex_match:
                rssi = int(regex_match.group(1))
                snr = int(regex_match.group(2))
                logger.debug(f"Received rssi: {rssi}, snr: {snr}")
            else:
                logger.debug(f"Received regex failed: {self._serial_buffer}")

            return success()
        return fail()

    def get_buffer(self) -> bytes:
        return self._buffer

    def _send_command(self, command: bytes, expected: bytes) -> bool:
        self._flush_serial()
        for i in range(self._num_retries):
            self._lora_serial.write(command)
            if not self._expected(expected):
                continue
            return True
        return False

    def _expected(self, expected: bytes) -> bool:
        START_TIME: float = time.time()
        self._serial_buffer = b""
        while len(self._serial_buffer) < len(expected) or self._serial_buffer[-len(expected):] != expected:
            # check if timeout reached
            if (self._lora_serial.in_waiting == 0) and (time.time() - START_TIME >= self._timeout): # `self._lora_serial.in_waiting == 0` so that we don't "cut-off" in the middle of receiving data
                return False
            
            # check if byte available to read (so that we don't block)
            if self._lora_serial.in_waiting:
                self._serial_buffer += self._lora_serial.read(1)
        return True
    
    def _read_blocking(self) -> int:
        START_TIME: float = time.time()
        while not self._lora_serial.in_waiting:
            if time.time() - START_TIME >= self._timeout:
                return -1
        return int(self._lora_serial.read(1)[0])
    
    def _flush_serial(self) -> None:
        if not self._lora_serial.in_waiting: # prevent blocking
            return
        self._lora_serial.read_all()
    
if __name__ == "__main__":
    LORA_COM_PORT: Literal['COM7'] = 'COM9'
    LORA_BAUDRATE: Literal[9600] = 9600
    # LoRa config
    LORA_TIMEOUT: float = 10.0
    LORA_NUM_RETRIES: Literal[5] = 5
    LORA_FREQ: Literal[915000000] = 915000000  # 915 MHz
    LORA_DATA_RATE: DataRate = DataRate.SF12
    LORA_BANDWIDTH: Bandwidth = Bandwidth.BANDWIDTH_125_KHZ
    LORA_CODE_RATE: CodeRate = CodeRate.RATE_4_BY_5
    LORA_TX_POWER: Literal[22] = 22
    LORA_LNA: LNA = LNA.ON
    LORA_LOW_DR_OPT: LowDrOpt = LowDrOpt.AUTO

    lora_serial: serial.Serial = serial.Serial(port=LORA_COM_PORT, baudrate=LORA_BAUDRATE)
    lora: LoRa = LoRa(lora_serial=lora_serial, timeout=LORA_TIMEOUT, num_retries=LORA_NUM_RETRIES, freq=LORA_FREQ, data_rate=LORA_DATA_RATE, bandwidth=LORA_BANDWIDTH, code_rate=LORA_CODE_RATE, tx_power=LORA_TX_POWER, lna=LORA_LNA, low_dr_opt=LORA_LOW_DR_OPT)
    print(f"LoRa begin: {lora.begin()}")
    print(f"LoRa RX: {lora.set_state(LoRaState.RX)}")
    # DATA: bytes = b"hello world"
    DATA: bytes = b"A" * 255
    print(f"Sending data: {lora.send(DATA)}")

    while True:
        if lora.recv():
            print(f"Data received: {lora.get_buffer().decode()}")
            time.sleep(0.4)
            print(f"Sending data: {lora.send(lora.get_buffer())}")

File: lib/LoRa/LoRa.cpp
#include "LoRa.h"

// source: https://stackoverflow.com/questions/25890784/computing-length-of-a-c-string-at-compile-time-is-this-really-a-constexpr
static int constexpr length(const char* str) {
    return *str ? 1 + length(str + 1) : 0;
}

LoRa::LoRa(LoRaSerial* lora_serial, const uint16_t timeout, const uint8_t num_retries, const uint32_t freq, const DataRate data_rate, const Bandwidth bandwidth, const CodeRate code_rate, const uint8_t tx_power, const LNA lna, const LowDrOpt low_dr_opt) : lora_serial_(lora_serial), timeout_(timeout), num_retries_(num_retries), freq_(freq), data_rate_(data_rate), bandwidth_(bandwidth), code_rate_(code_rate), tx_power_(tx_power), lna_(lna), low_dr_opt_(low_dr_opt) {}

bool LoRa::begin() {
    this->lora_serial_->flush_buffer();
    if (!this->reset()) {
        return false;
    }
    if (!this->configure()) {
        return false;
    }
    this->state_ = LoRaState::IDLE;
    return true;
}

bool LoRa::reset() {
    const char command_buffer[] = AT_RESET "\r\n";
    const char expected_buffer[] = "AT? to list all available functions\r\n";
    return this->send_command_(command_buffer, strlen(command_buffer), expected_buffer, strlen(expected_buffer));
}

static char* ultoa_internal(char* buffer, size_t buf_size, unsigned long val) {
    char* ptr = &buffer[buf_size - 1];
    *ptr = '\0';
    do {
        *--ptr = '0' + (val % 10);
        val /= 10;
    } while (val > 0);
    return ptr;
}

static char* utoa_internal(char* buffer, size_t buf_size, unsigned int val) {
    char* ptr = &buffer[buf_size - 1];
    *ptr = '\0';
    do {
        *--ptr = '0' + (val % 10);
        val /= 10;
    } while (val > 0);
    return ptr;
}

// copy string and return new end pointer (different to normal `strcpy` sort of like `strcat`)
static char* strcpy_internal(char* dest, const char* src) {
    while (*src) {
        *dest++ = *src++;
    }
    return dest;
}

bool LoRa::configure() {
    char command_buffer[length(AT_CCONF) + 40];
    // snprintf(command_buffer, ARR_SIZE(command_buffer), AT_CCONF "=%lu:%u:%u:%u:4/%u:%u:%u\r\n", this->freq_, this->tx_power_, (uint8_t)this->bandwidth_, (uint8_t)this->data_rate_, (uint8_t)this->code_rate_, (uint8_t)this->lna_, (uint8_t)this->low_dr_opt_);
    char* ptr = command_buffer;
    
    // Copy the AT command prefix
    ptr = strcpy_internal(ptr, AT_CCONF "=");
    
    // Convert frequency
    char freq_buf[12];  // Max 10 digits for 32-bit + null
    char* freq_ptr = ultoa_internal(freq_buf, sizeof(freq_buf), this->freq_);
    ptr = strcpy_internal(ptr, freq_ptr);
    *ptr++ = ':';
    
    // Convert and append other parameters
    char num_buf[6];  // Max 5 digits + null
    
    ptr = strcpy_internal(ptr, utoa_internal(num_buf, sizeof(num_buf), this->tx_power_));
    *ptr++ = ':';
    
    ptr = strcpy_internal(ptr, utoa_internal(num_buf, sizeof(num_buf), (uint8_t)this->bandwidth_));
    *ptr++ = ':';
    
    ptr = strcpy_internal(ptr, utoa_internal(num_buf, sizeof(num_buf), (uint8_t)this->data_rate_));
    *ptr++ = ':';
    
    *ptr++ = '4';
    *ptr++ = '/';
    
    ptr = strcpy_internal(ptr, utoa_internal(num_buf, sizeof(num_buf), (uint8_t)this->code_rate_));
    *ptr++ = ':';
    
    ptr = strcpy_internal(ptr, utoa_internal(num_buf, sizeof(num_buf), (uint8_t)this->lna_));
    *ptr++ = ':';
    
    ptr = strcpy_internal(ptr, utoa_internal(num_buf, sizeof(num_buf), (uint8_t)this->low_dr_opt_));
    
    // Add line ending
    *ptr++ = '\r';
    *ptr++ = '\n';
    *ptr = '\0';
    
    const char expected_buffer[] = "\r\nOK\r\n";
    return this->send_command_(command_buffer, ptr - command_buffer, expected_buffer, strlen(expected_buffer));
}

bool LoRa::send_command_(const char* command, const size_t command_length, const char* expected, const size_t expected_length) {
    this->lora_serial_->flush_buffer();
    for (uint8_t i = 0; i < this->num_retries_; ++i) {
        watchdog.refresh();
        this->lora_serial_->write(command, command_length);
        if (!this->lora_serial_->expected(expected, expected_length, serial_buffer_, LORA_SERIAL_BUFFER_SIZE, this->timeout_)) {
            continue;
        }
        return true;
    }
    return false;
}

static void byte_to_hex(uint8_t byte, char* out) {
    static const char hex_chars[] = "0123456789abcdef";
    out[0] = hex_chars[byte >> 4];
    out[1] = hex_chars[byte & 0x0F];
}

bool LoRa::send(const char* data, const size_t data_size) {
    if (data_size > LORA_MAX_SIZE) {
        return false;
    }

    DEBUG_LORA_PRINT("LoRa current state: "); DEBUG_LORA_PRINTLN((uint8_t)this->state_);

    const LoRaState saved_state = this->state_;
    if (!this->set_state(LoRaState::TX)) {
        this->set_state(saved_state);   
        return false;
    }

    constexpr size_t buffer_size = ARR_SIZE(this->buffer_);
    constexpr size_t tx_command_size = length(AT_CTX) + 1; // `+ 1` for the `"="`
    strncpy(this->buffer_, AT_CTX "=", buffer_size);
    
    // append data in 2-byte padded hex format with no separation
    size_t data_offset = tx_command_size;
    for (size_t i = 0; i < data_size; ++i) {
        if (data_offset + 2 >= buffer_size) {
            break;
        }
        // snprintf(buffer_ + data_offset, buffer_size - data_offset, "%02x", data[i]);
        byte_to_hex(data[i], buffer_ + data_offset);
        data_offset += 2;
    }

    strncpy(this->buffer_ + data_offset, "\r\n", buffer_size - data_offset);

    this->buffer_counter_ = data_size;
    const char expected_buffer[] = "OnTxDone\r\n";
    if (!this->send_command_(this->buffer_, data_offset + 2, expected_buffer, strlen(expected_buffer))) { // the `+ 2` is for the extra `"\r\n"`
        this->set_state(saved_state);
        return false;
    }

    this->set_state(saved_state); // ignore return value, we still successfully sent
    return true;
}

bool LoRa::set_state(const LoRaState updated_state) {
    switch (updated_state) {
        case LoRaState::RX:
            return enter_rx_mode_();
        case LoRaState::TX:
            return enter_tx_mode_();
        case LoRaState::IDLE:
            return enter_idle_mode_();
        case LoRaState::SLEEP:
            return enter_idle_mode_(); // `IDLE` is already low power
        default:
            return false;
    }
}

LoRaState LoRa::get_state() {
    return this->state_;
}

bool LoRa::enter_rx_mode_() {
    if (!configure()) {
        return false;
    }

    const char command_buffer[] = AT_CRX "\r\n";
    const char expected_buffer[] = "Receiving...\r\n";
    return this->send_command_(command_buffer, strlen(command_buffer), expected_buffer, strlen(expected_buffer));
}

bool LoRa::enter_tx_mode_() {
    if (!enter_idle_mode_()) {
        return false;
    }

    return configure();
}

bool LoRa::enter_idle_mode_() {
    const char command_buffer[] = AT_COFF "\r\n";
    const char expected_buffer[] = "Idle...\r\n";
    return this->send_command_(command_buffer, strlen(command_buffer), expected_buffer, strlen(expected_buffer));
}

uint8_t hex_char_to_byte(char c) {
    if (c >= '0' && c <= '9') {
        return c - '0';
    } else if (c >= 'a' && c <= 'f') {
        return c - 'a' + 10;
    } else if (c >= 'A' && c <= 'F') {
        return c - 'A' + 10;
    }
    // invalid
    return (uint8_t)-1;
}

bool LoRa::recv() {
    // ideally, we should already be in RX mode, so that if recv is called a previous message can be read from the serial buffer
    const LoRaState saved_state = this->state_;
    if (!this->set_state(LoRaState::RX)) {
        goto fail;
    }
    for (uint8_t i = 0; i < this->num_retries_; ++i) {
        watchdog.refresh();
        buffer_counter_ = 0;
        const char expected_buffer[] = "Recv:\r\n";
        if (!this->lora_serial_->expected(expected_buffer, strlen(expected_buffer), serial_buffer_, LORA_SERIAL_BUFFER_SIZE, this->timeout_)) {
            continue;
        }
        int c1 = this->lora_serial_->read_blocking(timeout_);
        if (c1 == -1) {
            goto fail;
        }
        int c2 = this->lora_serial_->read_blocking(timeout_);
        if (c2 == -1) {
            goto fail;
        }
        while (c1 != '\r') {
            watchdog.refresh();
            if (buffer_counter_ >= LORA_MAX_SIZE) {
                goto fail;
            }
            const uint8_t c1_byte = hex_char_to_byte(c1);
            const uint8_t c2_byte = hex_char_to_byte(c2);
            if ((c1_byte == (uint8_t)-1) || (c2_byte == (uint8_t)-1)) { // invalid hex formats
                goto fail;
            }
            this->buffer_[buffer_counter_] = (c1_byte << 4) | c2_byte;
            buffer_counter_++;
            if (this->lora_serial_->read_blocking(timeout_) != ' ') { // flush ' ' char
                goto fail;
            }
            c1 = this->lora_serial_->read_blocking(timeout_);
            if (c1 == -1) {
                goto fail;
            }
            c2 = this->lora_serial_->read_blocking(timeout_);
            if (c2 == -1) {
                goto fail;
            }
        }
        // the rx ends with `"Data end\r\nrssi = {rssi} dBm, snr = {snr} dB\r\n"`
        const char expected_buffer_2[] = "Data end\r\nrssi = ";
        if (!this->lora_serial_->expected(expected_buffer_2, strlen(expected_buffer_2), serial_buffer_, LORA_SERIAL_BUFFER_SIZE, this->timeout_)) {
            goto fail;
        }
        const char expected_buffer_3[] = " dB\r\n";
        if (!this->lora_serial_->expected(expected_buffer_3, strlen(expected_buffer_3), serial_buffer_, LORA_SERIAL_BUFFER_SIZE, this->timeout_)) {
            goto fail;
        }
        #ifdef DEBUG
        int rssi = 0, snr = 0;
        sscanf(serial_buffer_, "%d dBm, snr = %d dB\r\n", &rssi, &snr);
        #endif
        DEBUG_LORA_PRINT("Received rssi: ");
        DEBUG_LORA_PRINT(rssi);
        DEBUG_LORA_PRINT(", snr: ");
        DEBUG_LORA_PRINTLN(snr);
        this->set_state(saved_state);
        return true;
    }
    fail:
    this->set_state(saved_state);   
    return false;
}

const char* LoRa::get_buffer() {
    return this->buffer_;
}

uint8_t LoRa::get_buffer_len() {
    return this->buffer_counter_;
}